src/Stan/Hie/Match.hs

```diff
@@ -66,15 +66,12 @@ data Pattern
 
 {- | Matching function that searches the array of types recursively.
 -}
-hieMatchType
+hieMatchPattern
     :: TypeIndex   -- ^ Index of the current expression type
     -> Pattern  -- ^ Our search query
     -> Array TypeIndex HieTypeFlat  -- ^ Array of all types in HIE file
```

Comment:
Here is just a thought I had: what if we reorder arguments in this function to the following order:

```haskell
hieMatchPattern :: Array TypeIndex -> Pattern -> TypeIndex -> Bool
```

With the new order you whould be able to use partially applied `hieMatchPattern` and rewrite things like `any (\i -> hieMatchPattern i pat hie_types)` to `any (hieMatchPattern hie_types pat)`. Also, probably, with some extra internal functions, you can avoid passing `arr` as an argument inside internal recursive functions, but that's a minor suggestion, maybe not worth it.